### DISCUSSION TITLE
Inertial scrolling/Kinetic scrolling is missing in the Linux version

### DISCUSSION THREAD

**Discussion by westerlind** [ghostty-org/ghostty#4193] [Feature Requests, Ideas]
Inertial scrolling is missing in the Linux version, while being available in the macOS version (as almost all apps there of course).

This is not as obvious of a feature on Linux as it is on macOS, but it is the default behaviour in Gnome these days, so with the focus on being GTK/Adwaita native, I think it does make a lot of sense to make this available at least as an option.

One example of a Gnome centric terminal that does support this is Ptyxis (which is the default terminal on Fedora since v41), and to me, that feature alone makes the general user experience of Ptyxis nicer than Ghostty at the moment, even though Ghostty is of course superior in many other ways.

Having been primarily a Mac user for many years, I have come to love the inertial scrolling behaviour, and I always strongly miss it where it isn't available.

**Comment by mitchellh** [comment]
I didn't realize GTK supported this. Can you point to any documentation on how we get inertial scrolling events? We manage inertial scrolling in the cross-platform libghostty core so it should just be a matter of sending libghostty the proper events from the GTK apprt. 

Note that Ptyxis is GPL so do NOT link to any of its source code or use it to PR anything.

**Comment by jcollie** [comment]
A quick google turns this up: https://docs.gtk.org/gtk4/class.ScrolledWindow.html#touch-support

Not sure if that's what ptyxis is using.

**Reply by taoky** [reply]
Ptyxis uses VTE, and it sets enable-fallback-scrolling as false (<https://gnome.pages.gitlab.gnome.org/vte/gtk4/property.Terminal.enable-fallback-scrolling.html>) and scroll-unit-is-pixels as true (<https://gnome.pages.gitlab.gnome.org/vte/gtk4/property.Terminal.scroll-unit-is-pixels.html>) to support inertial scrolling.

(This could be tested with GTK inspector (`GTK_DEBUG=interactive`))

**Comment by westerlind** [comment]
GTK seems to call it kinetic scrolling rather than intertial scrolling, but it's the same thing.

https://docs.gtk.org/gtk4/property.ScrolledWindow.kinetic-scrolling.html

**Reply by jcollie** [reply]
Given that this needs a `ScrolledWindow` we probably might as well implement a scroll bar as well if someone implements this.

**Reply by westerlind** [reply]
I would for sure appreciate a scrollbar too, but I guess that doing it this way might have other downsides that we don't wont, like not being able to have inertial scrolling for terminal apps for example, and I guess it would change the structure of things quite a bit.

One alternative way of doing it would be to calculate the intertial scrolling in Ghostty instead (which I guess is what GTK simply helps apps doing when they use ScrolledWindow, as it doesn't seem to be done on the OS level the way it is in macOS), and base it on something like the following simple description of how Apple calculated the inertial scrolling in an unofficial Javascript library they seem to have made a long time ago. Not necessarily the exact same numbers as what GTK is using, so the numbers should possibly be slightly different, but the calculation needed to make it work well does not seem to be all that complex based on this anyway.

https://stackoverflow.com/questions/38619717/need-help-dissecting-and-recreating-the-perfect-scroll-easing-based-on-pastrykit

> On touch end Apple would get momentum by dividing number of pixels that the user had swiped, and time that the user has swiped for. If the number of pixels was less than 10 or time was less than 0.5, momentum would be clamped to zero.
Anyways, once the momentum (speed) was known to us, they would multiply it by 0.95 in every frame, and then move the screen by that much.

**Comment by westerlind** [comment]
Is it maybe reasonable to implement this in a similar way as the proof-of-concept for a solution for a scrollbar on macOS here: https://github.com/ghostty-org/ghostty/issues/111#issuecomment-2568359738 (maybe that was what you meant previously @jcollie?)

My understanding is that this would mean creating a ScrolledWindow that hosts the actual Ghostty terminal view, which "thinks" that it's scrolling the terminal content even though it isn't actually doing that, but because it behaves the same way weather it actually scrolls the content or not, that can be used to get the kinetic-scrolling events, and probably also to render the scrollbars.

My understanding is that Ptyxis is doing this by actually letting ScrolledWindow scroll the terminal content, and then they get both the scrollbar and the kinetic scrolling without having to do all that much more, but I guess that it would probably be possible to do it as described above too, and to me that seems like a better fit for Ghostty?

If I understand https://github.com/ghostty-org/ghostty/discussions/2524#discussioncomment-11087198 and https://github.com/ghostty-org/ghostty/issues/111 right, the plan is to do it this way on macOS for the scrollbar, so then it would make sense to structure it the same way on GTK too for the scrollbar and kinetic scrolling.

**Comment by paolomainardi** [comment]
This feature would be handy. Scrolling a terminal in Linux with a trackpad without kinetic scrolling is very tough, and it would make Ghostty the only modern, cross-platform, GPU-accelerated terminal to support this on Linux. I'm not even close to being an expert in Zig or low-level terminal programming. Otherwise, I would be very keen to help.

**Reply by pluiedev** [reply]
Please reread our discussion rules - no bump comments please!

**Comment by infactai** [comment]
I'm on the latest "main" commits as of a couple hours ago to remove the memory issue, which is in 1.2.3.

One of the things I'm looking to disable is actually the what I call "Physics based scrolling" in Ghostty. What you guys are calling "Inertial scrolling/Kinetic scrolling"

Ghostty is the only application on my system which is doing this. 

My Gnome "does" have this effect, but let's say the "resistance to the inertia" is very high and it's barely noticeable (aka, when I stop scrolling the window stops scrolling near "immediately")

But with ghostty it seems someone has "oiled" up the scroll bar and it goes for quite a while. Additionally you can "flick" it and get heavy inertia and it goes very far.

For me, I don't use a mouse or keyboard, but instead of the track point, but I don't think that would make a difference. 

Ghostty is the only application which behaves like this. 

If this were an FPS game, I can't "aim" when using Ghostty. It's like it has a different "mouse accel + sensitivity" vs ever other application/window in my system and it's jarring.

**Comment by luca-schlecker** [comment]
## Overview

I have been experimentally implementing kinetic scrolling for Ghostty. I have identified two approaches for this, the trade-offs of which will be detailed in the following. I'd like some input from the maintainers (@mitchellh) on these, especially regarding which method they'd like to move forward with.

https://github.com/user-attachments/assets/689aac34-0e7d-4c75-b960-ec6a465268f5

GTK has implemented support for kinetic scrolling in their `Gtk.ScrolledWindow` widget. A consumer can be notified about the scroll position by subscribing to the `changed` signal of the according vertical `Gtk.Adjustment`. Ghostty is already using said widget and is already handling said signal to set the viewport scroll position when clicking on the scrollbar. However, kinetic scrolling is not affected by this as Ghostty registers a `Gtk.EventControllerScroll`, which consumes the scroll events before they are bubbled up to the `Gtk.ScrolledWindow`. As a result, the window never knows about the user having scrolled and never triggers any kinetic scrolling updates.

The `vadjValueChanged` callback handling the `Gtk.Adjustment` updates directly scrolls the viewport without calling the `surface.scrollCallback` method. As a consequence, the scroll updates by `vadjValueChanged` will not handle scrolling for mouse-reporting mode (e. g. TUI applications).

## Approaches

### Approach 1

The first approach makes use of the `Gtk.ScrolledWindow` logic to kinetically scroll the viewport. As far as I can tell, API limitations prevent us from properly making use of the `Gtk.Adjustment` updates for mouse-mode reporting. These events are only fired if the window is actually scrollable, which it isn't when opening a TUI.

Either:
**1a) Only the viewport is scrolled kinetically and TUIs will **not** be scrolled kinetically.**
**1b) Hack support for kinetic mouse-reporting by setting the scrollable area of the `Gtk.ScrolledWindow` to something bigger and constantly resetting the scroll-position to the center (similar to mouse capturing in games).**

This would be consistent with GTK and I expect it to be less overall code. However, the center-fixing hack to get kinetic mouse-reporting to work seems weird and I'd probably prefer _Approach 2_ over _1b)_.

### Approach 2

Implement kinetic scrolling manually, separate from the `Gtk.ScrolledWindow`. This provides full control and is partially assisted by `Gtk.EventControllerScroll`, which would provide velocity values for kinetic scrolling. Only the deceleration part would have to be implemented (for example a timer multiplying the velocity by 0.95 every 1ms or something, this can be arbitrary).

This would be more work to implement but enables us to choose own deceleration curves (if that is even desired) and kinetic mouse-reporting.

---

I'd happily volunteer to implementing any of the above.

**Comment by ezterry** [comment]
My LLM generated version for use on a mini laptop with very bad non touch mouse support I generally skipped the kinetic part, so no flick support on the scroll just mimicking middle wheel .  To me this seems right I usually want more precision not large scrolls.

My tests are on my own linux mint/ubuntu noble package.

This includes some scrolling history, using the scroll in some TUI apps as well  (For me the TUI is more important than normal scroll-back)

(created by Claude 4.6 reads sane but I've not really looked at other integration so really just for reference/expermiments)
```diff
diff --git a/src/apprt/gtk/class/surface.zig b/src/apprt/gtk/class/surface.zig
index 344bf8f21..3c305700a 100644
--- a/src/apprt/gtk/class/surface.zig
+++ b/src/apprt/gtk/class/surface.zig
@@ -518,6 +518,12 @@ pub const Surface = extern struct {
         /// True when we have a precision scroll in progress
         precision_scroll: bool = false,
 
+        /// True when a touchscreen drag-to-scroll is in progress
+        touch_drag_active: bool = false,
+
+        /// Accumulated touch drag Y offset for computing delta in drag-update
+        touch_drag_offset_y: f64 = 0,
+
         /// True when the child has exited.
         child_exited: bool = false,
 
@@ -2370,6 +2376,54 @@ pub const Surface = extern struct {
         return 1;
     }
 
+    fn gdTouchDragBegin(
+        _: *gtk.GestureDrag,
+        _: f64,
+        _: f64,
+        self: *Self,
+    ) callconv(.c) void {
+        const priv = self.private();
+        priv.touch_drag_active = true;
+        priv.touch_drag_offset_y = 0;
+    }
+
+    fn gdTouchDragUpdate(
+        _: *gtk.GestureDrag,
+        _: f64,
+        offset_y: f64,
+        self: *Self,
+    ) callconv(.c) void {
+        const priv = self.private();
+        if (!priv.touch_drag_active) return;
+
+        const surface = priv.core_surface orelse return;
+
+        const delta_y = offset_y - priv.touch_drag_offset_y;
+        priv.touch_drag_offset_y = offset_y;
+
+        if (delta_y == 0) return;
+
+        const scaled = self.scaledCoordinates(0, delta_y);
+        surface.scrollCallback(
+            0,
+            scaled.y,
+            .{ .precision = true },
+        ) catch |err| {
+            log.warn("error in touch drag scroll callback err={}", .{err});
+        };
+    }
+
+    fn gdTouchDragEnd(
+        _: *gtk.GestureDrag,
+        _: f64,
+        _: f64,
+        self: *Self,
+    ) callconv(.c) void {
+        const priv = self.private();
+        priv.touch_drag_active = false;
+        priv.touch_drag_offset_y = 0;
+    }
+
     fn imPreeditStart(
         _: *gtk.IMMulticontext,
         self: *Self,
@@ -2877,6 +2931,9 @@ pub const Surface = extern struct {
             class.bindTemplateCallback("scroll", &ecMouseScroll);
             class.bindTemplateCallback("scroll_begin", &ecMouseScrollPrecisionBegin);
             class.bindTemplateCallback("scroll_end", &ecMouseScrollPrecisionEnd);
+            class.bindTemplateCallback("touch_drag_begin", &gdTouchDragBegin);
+            class.bindTemplateCallback("touch_drag_update", &gdTouchDragUpdate);
+            class.bindTemplateCallback("touch_drag_end", &gdTouchDragEnd);
             class.bindTemplateCallback("drop", &dtDrop);
             class.bindTemplateCallback("gl_realize", &glareaRealize);
             class.bindTemplateCallback("gl_unrealize", &glareaUnrealize);
diff --git a/src/apprt/gtk/ui/1.2/surface.blp b/src/apprt/gtk/ui/1.2/surface.blp
index 7ed78ecb3..76d255bb5 100644
--- a/src/apprt/gtk/ui/1.2/surface.blp
+++ b/src/apprt/gtk/ui/1.2/surface.blp
@@ -158,6 +158,13 @@ Overlay terminal_page {
     button: 0;
   }
 
+  GestureDrag {
+    drag-begin => $touch_drag_begin();
+    drag-update => $touch_drag_update();
+    drag-end => $touch_drag_end();
+    touch-only: true;
+  }
+
   DropTarget drop_target {
     drop => $drop();
     actions: copy;
```

**Reply by luca-schlecker** [reply]
This does not affect two-finger trackpad scroll. It would be interesting for touch-devices though.

**Comment by barneysspeedshop** [comment]
I have inertial scrolling but I don't want it. How do I turn it off???
